design/deploy.md

```diff
@@ -0,0 +1,98 @@
+# Deploy
+
+A proposal to add support for hosting service deployment to the Polymer CLI.
+
+## Goals
+- Provide new users with a fast way to get a project hosted online.
+- Provide advanced users with an initial bootstrap for more advanced deployments.
+- Demonstrate the [PRPL](https://www.polymer-project.org/2.0/toolbox/server) pattern where HTTP/2 Server Push is available.
+
+## Non-goals
+- Do not try to support every possible deployment environment and configuration, or even a lot of them. Just enough to get users started.
+
+## Design
+
+- Add an `--environment` flag.
+  - Every command knows that it's being run in the context of this environment.
+  - Under the `default` environment, all commands operate as they do today.
+  - Under other environments:
+     - `build` now outputs the boilerplate needed to deploy the project to a given service, in addition to the traditional app build output.
+     - `serve` runs an appropriate HTTP server that emulates the configured hosting environment, instead of polyserve.
+
+- Add an `environments` section to `polymer.json`.
```

Comment:
There are two concepts. We should iterate on names.

The first concept is what I was calling an "environment". It's a specific user deployment configuration (e.g. my particular app engine deploy target). I imagined a polymer.json can have many of these defined in the "environments" section, similar to builds. Maybe your project can be deployed to both GitHub Pages and App Engine. Or maybe you have two App Engine targets, e.g. testing and prod. The --environment flag would pick from one of these configurations. We could also call this an "environment configuration"?

The second concept is what I was calling an "environment type". It's the higher level concept. Each environment would specify one of these types. This doc describes two such types so far: App Engine and GitHub Pages. We could also call this "environment" (and rename the first concept), "environment plugin", or "environment template"?